### DISCUSSION TITLE
SHA256 vs Blake3 for file hashing

### DISCUSSION THREAD

**Discussion by guri87-byte** [paulmillr/noble-hashes#77] [Q&A] [answered]
Currently I use sha256 hash in my product for file content hashing, recently moved to noble-sha256 from alternative JS library and got around 100% improvement in speed.
afaik, Blake3 is much more performant compared to SHA family, I tested locally with b3sum (rust implementation) and any sha family hash, b3sum is way faster.
But when I tried same with the library and compared sha256 and b3 speed, the sha256 was constantly 2x faster.
What can be possible reason?

https://github.com/BLAKE3-team/BLAKE3
![](https://github.com/BLAKE3-team/BLAKE3/raw/master/media/speed.svg)

**Comment by paulmillr** [comment, marked as answer]
1. We can’t use 64-bit arrays, they are slow: we are limited to 32b
2. We can’t use threads and SIMD (parallelization)

On its own, blake3 is slower than sha2. It can achieve these performance numbers via clever tricks, which are not always available.

**Comment by flawrist** [comment]
Online implementations of BLAKE3 which I found satisfactory:
* https://toolkitbay.com/tkb/tool/BLAKE3
* https://connor4312.github.io/blake3/
